File: Data Structure/Hash/Medium/LET_no2131_Longest Palindrome by Concatenating Two Letter Words.py
class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        memo = [[0 for _ in range(26)] for _ in range(26)]
        res = 0
        for w in words:
            a, b = ord(w[0]) - ord('a'), ord(w[1]) - ord('a')
            if memo[b][a] > 0:
                res += 4
                memo[b][a] -= 1
            else:
                memo[a][b] += 1
        
        for i in range(26):
            if memo[i][i] > 0:
                res += 2
                break
        
        return res

File: Data Structure/Hash/Medium/LET_no2131_Longest Palindrome by Concatenating Two Letter Words.cpp
class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        vector<vector<int>> memo(26, vector<int>(26, 0));
        int res = 0;
        for (string &w : words) {
            int a = w[0] - 'a';
            int b = w[1] - 'a';
            
            if (memo[b][a] > 0) {
                res += 4;
                memo[b][a]--;
            }
            else
                memo[a][b]++;
        }
        
        for (int i = 0; i < 26; i++) {
            if (memo[i][i] > 0) {
                res += 2;
                break;
            }
        }
        
        return res;
    }
};